### PR TITLE
Footprint improvements

### PR DESCRIPTION
**Pull request by wirthi**
* More laziness when creating child nodes
* avoid creating 0-element arrays (Truffle node cloning is re-creating some of them, though)

### CODE REVIEW

File: src/main/java/org/truffleruby/parser/BodyTranslator.java [hunk 1]
```diff
@@ -779,7 +779,7 @@ protected ArgumentsAndBlockTranslation translateArgumentsAndBlock(SourceIndexLen
             throw new UnsupportedOperationException("Unknown argument node type: " + argsNode.getClass());
         }
 
-        final RubyNode[] argumentsTranslated = new RubyNode[arguments.length];
+        final RubyNode[] argumentsTranslated = arguments.length == 0 ? RubyNode.EMPTY_ARRAY : new RubyNode[arguments.length];
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/BodyTranslator.java]: Could you add a helper method like `createRubyNodeArray(int size)` to share the logic?

File: src/main/java/org/truffleruby/core/RaiseIfFrozenNode.java [hunk 2]
```diff
@@ -32,11 +33,19 @@ public RaiseIfFrozenNode(RubyNode child) {
...
-        if (isFrozenNode.executeIsFrozen(result)) {
+        if (getIsFrozen().executeIsFrozen(result)) {
             errorProfile.enter();
             throw new RaiseException(coreExceptions().frozenError(result, this));
         }
 
         return result;
     }
+
+    private IsFrozenNode getIsFrozen() {
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/RaiseIfFrozenNode.java]: This should be `isFrozen` and include the call to `executeIsFrozen(result)`, so it's nicer at the call site.
(From https://github.com/oracle/truffleruby/blob/master/doc/contributor/code-patterns.md#the-lazy-pattern)

File: src/main/java/org/truffleruby/core/cast/ToSNode.java [hunk 3]
```diff
@@ -33,20 +33,27 @@ protected DynamicObject kernelToS(Object object) {
...
+
     @Specialization(guards = "isRubyString(string)")
     public DynamicObject toS(DynamicObject string) {
         return string;
     }
 
     @Specialization(guards = "!isRubyString(object)")
     public DynamicObject toSFallback(VirtualFrame frame, Object object) {
-        final Object value = callToSNode.call(frame, object, "to_s");
+        final Object value = getCallToSNode().call(frame, object, "to_s");
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/cast/ToSNode.java]: Could it be created in `@Cached` or is that worse for footprint?

File: src/main/java/org/truffleruby/language/constants/ReadConstantNode.java [hunk 4]
```diff
@@ -35,9 +36,25 @@ public ReadConstantNode(RubyNode moduleNode, String name) {
...
+        }
+        return getConstantNode;
+    }
+
+    private LookupConstantNode getLookupConstantNode() {
+        if (lookupConstantNode == null) {
+            CompilerDirectives.transferToInterpreterAndInvalidate();
+            lookupConstantNode = insert(LookupConstantNodeGen.create(false, false));
+        }
+        return lookupConstantNode;
```
Comment (eregon) [inline src/main/java/org/truffleruby/language/constants/ReadConstantNode.java]: Same comment regarding including the `execute` call inside the lazy pattern method.

File: src/main/java/org/truffleruby/core/cast/ToStrNode.java [hunk 5]
```diff
@@ -43,7 +44,7 @@ public DynamicObject coerceObject(VirtualFrame frame, Object object,
             @Cached("create()") BranchProfile errorProfile) {
         final Object coerced;
         try {
-            coerced = toStrNode.call(frame, object, "to_str");
+            coerced = getToStrNode().call(frame, object, "to_str");
```
Comment (eregon) [inline src/main/java/org/truffleruby/core/cast/ToStrNode.java]: Can it be created in `@Cached` or is it better like this for footprint?

File: src/main/java/org/truffleruby/core/RaiseIfFrozenNode.java (hunk 2 above)
Comment (wirthi) [reply src/main/java/org/truffleruby/core/RaiseIfFrozenNode.java]: thanks, fixed.

File: src/main/java/org/truffleruby/core/cast/ToSNode.java (hunk 3 above)
Comment (wirthi) [reply src/main/java/org/truffleruby/core/cast/ToSNode.java]: fixed

File: src/main/java/org/truffleruby/core/cast/ToStrNode.java (hunk 5 above)
Comment (wirthi) [reply src/main/java/org/truffleruby/core/cast/ToStrNode.java]: fixed

File: src/main/java/org/truffleruby/language/constants/ReadConstantNode.java (hunk 4 above)
Comment (wirthi) [reply src/main/java/org/truffleruby/language/constants/ReadConstantNode.java]: true, but here it comes at the cost of some code duplication, as `executeGetConstant` also requires `lookupConstant`.

File: src/main/java/org/truffleruby/parser/BodyTranslator.java (hunk 1 above)
Comment (wirthi) [reply src/main/java/org/truffleruby/parser/BodyTranslator.java]: fixed, create `RubyNode.createArray(int size)`

File: src/main/java/org/truffleruby/language/constants/ReadConstantNode.java (hunk 4 above)
Comment (eregon) [reply src/main/java/org/truffleruby/language/constants/ReadConstantNode.java]: Ah, right. I missed `executeGetConstant()` takes a LookupConstantNode as argument as well. Good as-is then.

File: src/main/java/org/truffleruby/parser/BodyTranslator.java [hunk 6]
```diff
@@ -3067,7 +3067,7 @@ public RubyNode visitStarNode(StarParseNode star) {
     }
 
     protected RubyNode initFlipFlopStates(SourceIndexLength sourceSection) {
-        final RubyNode[] initNodes = new RubyNode[environment.getFlipFlopStates().size()];
+        final RubyNode[] initNodes = environment.getFlipFlopStates().size() == 0 ? RubyNode.EMPTY_ARRAY : new RubyNode[environment.getFlipFlopStates().size()];
```
Comment (eregon) [inline src/main/java/org/truffleruby/parser/BodyTranslator.java]: This one can also use `RubyNode.createArray()` I think.

**Pull request merged**